pywhy_graphs/algorithms/generic.py

```diff
@@ -333,3 +334,259 @@ def _single_shortest_path_early_stop(G, firstlevel, paths, cutoff, join, valid_p
+    cur_node : node
+        The current node.
+    prev_node : node
+        The previous node in the path.
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path which is empty if the bool is false.
+    """
+    path_exists = False
+    path = []
+    visited.add(cur_node)
+    neighbors = G.neighbors(cur_node)
+
+    if cur_node is node_y:
+        return (True, [node_y])
+
+    for elem in neighbors:
+        if elem in visited:
+            continue
+
+        else:
+            # If the current node is a collider, check that it is either an
+            # ancestor of X, Y or any element of S or that it is
+            # the destination node itself.
+            if (
+                _is_collider(G, prev_node, cur_node, elem)
+                and (cur_node not in all_ancestors)
+                and (cur_node not in S)
+                and (cur_node is not node_y)
+            ):
+                continue
+
+            # If the current node is not a collider, check that it is
+            # either in L or the destination node itself.
+
+            elif (
+                not _is_collider(G, prev_node, cur_node, elem)
+                and (cur_node not in L)
+                and (cur_node is not node_y)
+            ):
+                continue
+
+            # if it is a valid node and not the destination node,
+            # check if it has a path to the destination node
+
+            path_exists, temp_path = _shortest_valid_path(
+                G, node_x, node_y, L, S, visited, all_ancestors, elem, cur_node
+            )
+
+            if path_exists:
+                path.append(cur_node)
+                path.extend(temp_path)
+                break
+
+    return (path_exists, path)
+
+
+def inducing_path(G, node_x: Node, node_y: Node, L: Set = None, S: Set = None):
+    """Checks if an inducing path exists between two nodes as defined in :footcite:`Zhang2008`.
+
+    Parameters
+    ----------
+    G : Graph
+        The graph.
+    node_x : node
+        The source node.
+    node_y : node
+        The destination node.
+    L : Set
+        Nodes that are ignored on the path. Defaults to an empty set. See Notes for details.
+    S:  Set
+        Nodes that are always conditioned on. Defaults to an empty set. See Notes for details.
+
+    Returns
+    -------
+    path : Tuple[bool, path]
+        A tuple containing a bool and a path if the bool is true, an empty list otherwise.
+
+    Notes
+    -----
+    An inducing path intuitively is a path between two non-adjacent nodes that
+    cannot be d-separated. Therefore, the path is always "active" regardless of
+    what variables we condition on. L contains all the non-colliders, these nodes
+    are ignored along the path. S contains nodes that are always conditioned on
+    (hence if the ancestors of colliders are in S, then those collider
+    paths are always "active").
```

Comment:
You need a References section at the end here. See some other docstrings.

Reply:
@adam2392 I added a references section and the build is still failing.

Reply:
Hmm weird. Perhaps try debugging locally. 

I can take a look this week / tmr.

Reply:
My local build has been giving me errors for a while. I will try debugging it. Another thing is the networkx tests are failing because apparently the support for Python 3.8 has been deprecated.

Reply:
It turns out you cannot have a `:footcite:`...`` pattern inside the header of the docstring, so I moved it to the description.